### PR TITLE
Add example of validation in custom field type

### PR DESCRIPTION
**Pull request by bjarnef**
image

image

image

### CODE REVIEW

**Comment by eshanrnh** [comment]
Thanks for the PR, @bjarnef 🙌 We will review this soon 🙂

**Comment by bjarnef** [comment]
I think @AndyButland may have some input and the "CompareField" setting currently just contain alias of the other field.

File: Add-ons/UmbracoForms/Developer/Extending/Adding-a-Fieldtype.md [hunk 1]
```diff
@@ -160,3 +160,106 @@ To reference the file you should override the `RenderView` property, e.g.:
...
+        {
+            return string.Empty;
+        }
+
+        return context.Request.HasFormContentType && context.Request.Form.Keys.Contains(field.Id.ToString())
+            ? context.Request.Form[field.Id.ToString()].ToString().Trim()
+            : string.Empty;
+    }
+
+    private static Field? GetPostField(Form form, string key) => form.AllFields.SingleOrDefault(f => f.Alias == key);
```
Comment (bjarnef) [inline Add-ons/UmbracoForms/Developer/Extending/Adding-a-Fieldtype.md:238]: @AndyButland I noticed these method in other examples, but maybe Forms should include these extensions methods?

E.g.
```
GetField(Form form, string key)
GetFieldValue(Form form, HttpContext context, string key)
```

**Review by AndyButland — COMMENTED** [review]
This is a nice solution, but I think it might be better as a blog post we link to (which I'd be happy to do so from the docs) rather than going in as official documentation.  There's a few reasons for this:

- There's a hard-coded reference to a form ID.  Whilst that might work for a particular scenario, it's not a very general approach that we can really recommend people do.
- You've got client-side validation working with jquery which is great, but not sure if it also works with `aspnet-client-validation`?  I think it would need to given people are often moving away from jquery now for something more lightweight.
- The compare field is missing most of the settings available on the text field.  And it would be an overhead in code and in docs to maintain this.

This is just an idea, but I think the best way to support this in Forms may be something you configure at the level of the form rather than the field.  So like you can add workflows for a form, perhaps we have some sort of means to add "form validation rules".  Validation on single fields would be set as is, but rules that span fields across the form would be set here.  So you could have "compare email = email", "end date > start date" etc. without having to have "compare" versions of many of the field types.

We would then need to have client side and server side validation of these rules as the form and/or page is submitted.

**Comment by bjarnef** [comment]
@AndyButland regarding the hard-coded reference to a form ID. Yes, I don't want it to be hard-coded, but I didn't found a way to get current form reference in field types (like in form template itself). Do field types have any access to the form context, e.g. get current form or just a `Guid` reference to the form?

Currently it is using `equal-to` which is available in jQuery validation, but may work without jQuery if the client side validation works using same attribute. The other field types use similar attributes like `data-val-required` , `data-val-regex` and `data-val-regex-pattern`. Not sure how it works without jQuery and jQuery validation? Does it just ignore these attributes or does it by coincidence have client side validations in `aspnet-client-validation` using same attributes?

Which specific settings are missing for Compare field? It is basically a copy / extended version of `FieldType.Textfield.cshtml`.

**Comment by AndyButland** [comment]
> Do field types have any access to the form context, e.g. get current form or just a Guid reference to the form?
From within the partial view, I don't believe so - they just have access to the `FieldViewModel`.

> Not sure how it works without jQuery and jQuery validation?  Does it just ignore these attributes or does it by coincidence have client side validations in aspnet-client-validation using same attributes?
I suspect the new library tries to follow the same conventions, as it's intended as a replacement for using jquery.  But it would need testing to confirm.

> Which specific settings are missing for Compare field? It is basically a copy / extended version of FieldType.Textfield.cshtm
I meant the settings field type - so autocomplete attribute, field type, placeholder etc.

**Comment by bjarnef** [comment]
@AndyButland okay, I am not sure how `FieldViewModel` is populated with data, but I guess since it has is populate with e.g. `Caption` and `Alias` of the field in the current form, this logic already have access to the form. So could all field types e.g. have a `FormId` property. Then it would be easier to find the form without hardcoding the `Guid` form reference.

Not sure if `aspnet-client-validation` is based on this, but in that case there's a `equalto` (compare) method as well:
https://github.com/haacked/aspnet-client-validation/blob/3d754e9613e84f689dee3fb77fbcbf08a84f731d/dist/aspnet-validation.js#L530

Yes, you're are right .. I added the settings to the view, but not to the field type. I tried originally to inherit from the `Textfield` provider, but this override this existing `Textfield` (Short answer) like this https://our.umbraco.com/documentation/Add-ons/UmbracoForms/Developer/Extending/Adding-a-Type#overriding-default-providers-in-umbraco-forms
But it was the intention to have same setting + addition settings.

Can we extend/inherit a field type from an existing field type e.g. `Textfield` instead of `Umbraco.Forms.Core.FieldType`? But then Umbraco Forms could e.g. later add new settings to `Textfield`, which may also require additional updates in the view of the custom field type.

**Comment by bjarnef** [comment]
@AndyButland I tried removing jQuery + jQuery validation and instead included `@Html.RenderUmbracoFormDependencies()` mentioned here: https://our.umbraco.com/documentation/Add-ons/umbracoforms/developer/Prepping-Frontend/

The compare client side validation doesn't quite seem to work, but the JS included /App_Plugins/UmbracoForms/Assets/aspnet-client-validation/dist/aspnet-validation.min.js does contain a `equalTo` provider method to compare fields.

**Review by sophie-neale — COMMENTED** [review]
Spotted a little duplication 👀

File: Add-ons/UmbracoForms/Developer/Extending/Adding-a-Fieldtype.md [hunk 2]
```diff
@@ -160,3 +160,106 @@ To reference the file you should override the `RenderView` property, e.g.:
...
+    [Setting("Field to compare",
+        Description = "Alias of field to compare.",
+        View = "textfield")]
+    public string FieldToCompare { get; set; }
+
+    /// <summary>
+    /// Gets or sets a value indicating whether the field label should be shown.
+    /// PreValues are a single element, a boolean indicating whether the default for the the checkbox is "checked".
+    /// </summary>
+    [Setting("Show Label", Description = "Indicate whether the the field's label should be shown when rendering the form.", View = "checkbox", PreValues = "true", DisplayOrder = 20)]
```
[the remark this answers is not in the document]
Comment (bjarnef) [reply Add-ons/UmbracoForms/Developer/Extending/Adding-a-Fieldtype.md:194]: Just copied from @AndyButland here: https://github.com/umbraco/Umbraco.Forms.Issues/issues/810#issuecomment-1182831403 ... so I wonder if the typo exists in Forms as well, but not from what I have noticed 😉

image

**Comment by sofietoft** [comment]
Hi all 👋 

What is the status of this PR?
It's been a while since we've seen any activity - @AndyButland suggests this would be better as a blog post, and we'd definitely be up for linking to it.
Is that something you'd be up for doing @bjarnef ? 😄 
It also looks like a related issue was solved 💪

**Comment by bjarnef** [comment]
@sofietoft I don't really write blog posts, but even anyone shared that I am not sure it is discoverable as the documentation. I think it could make sense to show a few examples / code snippets to extend CMS, Forms etc.

I think it is a typical example to add custom validation and often it probably depends on another form field like in this use-case.

**Comment by sofietoft** [comment]
Hi @bjarnef !

I've had a chat with @AndyButland about this, and we've agreed to not add this sample to our docs.
We do not feel that this sample represents the general best practice that we want the Umbraco docs to represent.

We do agree, though, that it's a sample that would definitely be valuable to other people. The optimal solution here would be to have the option to add this sample in the docs and then marking it as "community content".
Unfortunately, we do not have a good way of doing something like that at the moment 😞 I'll be having a chat with the DevRel team about it, as I think it would be really valuable having a place for these community contributions that don't quite fit our docs but still has a lot of value to them!

I'll be closing this PR for now though.
I hope you understand! Please do let me know if you have further questions 🙌 

And thanks again for putting in the work here - we really appreciate it!

**Pull request closed without merging**